src/util/pg.py

```diff
@@ -297,10 +298,10 @@ def explode_query_range(cr, query, table, alias=None, bucket_size=10000, prefix=
         # Still, since the query may only be valid if there is no split, we force the usage of `prefix` in the query to
         # validate its correctness and avoid scripts that pass the CI but fail in production.
         parallel_filter = "{alias}.id IS NOT NULL".format(alias=alias)
-        return [query.format(parallel_filter=parallel_filter)]
+        return [query.replace(PARALLEL_FILTER, parallel_filter)]
```

Comment:
If we go forward with this I would suggest you keep the previous behavior and replace `{{` -> `{` and `}}` -> `}`. That should avoid the need to adapt already existing queries.
Maybe:
```py
re.sub(r"(?<!\{)\{parallel_filter\}(?!\})", parallel_filter, query).replace("{{", "{").replace("}}", "}")
```

Reply:
> `re.sub(r"(?<!\{)\{parallel_filter\}(?!\})", parallel_filter, query)`

I had the exact same idea, however its only advantage wrt the regular `replace` is that it would allow queries to have an hard-coded instance of the literal `'{parallel_filter}'`. Not that it costs us much, but do we really need that? I think that reserving one special placeholder for that util is fair.

> .replace("{{", "{").replace("}}", "}")

Out of caution we can think of doing that. However, IIMO the upside of not having to introduce this now-totally-artificial, odd and legacy-code-duplicating side-effect are worth considering:

1. It would be less surprising for people writing upgrade scripts (especially first timers).
2. In the long term it would spare us some challenging reviews/bugfixes due to complex setups (e.g. `util.explode_execute(cr, cr.mogrify(util.format_query(cr, f"")))`, each of which would introduce an extra set of curly braces) with dynamically baked-in values. Not to mention that python 3.12 is upon us and will introduce even more f-string capabilities to deal with.
3. People would be able to do `UPDATE t SET c = '{'` to assign `"{"` just fine, but would have to do `UPDATE t SET c = '{{{'` to assign `"{{"`? Not to mention that we would need to engineer something more clever (a regex) to resolve higher-order issues:

```python
>>> "{{{{".replace("{{", "{")
'{{'
```

In summary, IMO it would be wiser to invest towards preventing issues with old queries and "forget" about this, than to try to keep it together, preserving this oddity forever.

Reply:
> its only advantage wrt the regular replace is that it would allow queries to have an hard-coded instance of the literal '{parallel_filter}'. Not that it costs us much, but do we really need that?

No, there is another advantage: you don't need to adapt the existing queries. Do you?

All the extra issues you mentioned are already present with the `format` version. What I aim to here is to not break existing code. You can put a warning about why we do the replace. Something like "this used to be done with str.format" :/

Reply:
For the record
```
>>> "{{{{".replace("{{", "{")
'{{'
>>> "{{{{".format()
'{{'
```
Nothing is broken here for existing code.

Reply:
> No, there is another advantage: you don't need to adapt the existing queries. Do you?

Well yes, but I couldn't find any query where `"{{parallel_filter}}"` was passed to `explode_query_range`. After all, why would one have wanted to? The only reason I can think of is the one I mentioned: to hardcode the literal placeholder in the query. And, unsurprisingly, - at least to the extent of my check - that has not happened in queries in our repo, yet.

> All the extra issues you mentioned are already present with the `format` version. 

You're right (well except for point 3: right now single hardcoded curly braces fail the format, instead of moving on) but that doesn't make them any less valid: those are issues we'd get rid of.

> What I aim to here is to not break existing code. 

Yeah, I understand that. Personally I checked all files (41 to be exact) matched by this: `rg -tpy "\{\{" src/upgrade/migrations`. Of course, there are many other ways one could get a pair of curly braces into a string (e.g. `"{"*2`), not sure how many (notice that only 4 queries needed adjustments after my search) we can expect of those, though.

> For the record
> 
> ```
> >>> "{{{{".replace("{{", "{")
> '{{'
> >>> "{{{{".format()
> '{{'
> ```
> 
> Nothing is broken here for existing code.

Indeed, my bad:

```python
>>> "{} {{{{}}}}".format("hi")
'hi {{}}'
```

Reply:
> Well yes, but I couldn't find any query where "{{parallel_filter}}"

This is a public repo. You will never find all usages out there.

What specifically you want to avoid? If you want to do a replace for `{parallel_filter}` I don't see any downside in ensuring you do not replace `{{parallel_filter}}`. Can you put a specific reason why do you prefer a bare replace over the regexp one? Besides the "now you have two problems" reason OFC.

Reply:
My main point is that if we try to preserve the curly braces reduction side-effect we improve this util, but only marginally. What happens to 
`util.explode_execute(cr, "UPDATE t SET c = '{"hi": {"hello": 3}}::jsonb'`, table="t")`? We remove one of the closing curly brace and and break the jsonb.

The age old way to solve this is by adding a flag `format_side_effect=True` to decide which behaviour to run in new instances and default to the old one.

Reply:
Well, I really don't see how we can keep compatibility with naive replace (even if regexp). Maybe a new flag `use_replace=False` to use replace when we know the query could be problematic will be better.